src/tasks/MonoTargetsTasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,461 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public abstract class EmitBundleBase : Microsoft.Build.Utilities.Task, ICancelableTask
+{
+    private CancellationTokenSource BuildTaskCancelled { get; } = new();
+
+    private Dictionary<string, string> resourceDataSymbolDictionary = new();
+
+    private Dictionary<string, string[]> resourcesForDataSymbolDictionary= new();
+
+    /// Truncate encoded hashes used in file names and symbols to 24 characters.
+    /// Represents a 128-bit hash output encoded in base64 format.
+    private const int MaxEncodedHashLength = 24;
+
+    /// Must have DestinationFile metadata, which is the output filename
+    /// Could have RegisteredName, otherwise it would be the filename.
+    /// RegisteredName should be prefixed with namespace in form of unix like path. For example: "/usr/share/zoneinfo/"
+    [Required]
+    public ITaskItem[] FilesToBundle { get; set; } = default!;
+
+    /// <summary>
+    /// The function to call before mono runtime initialization
+    /// in order to register the bundled resources in FilesToBundle
+    /// </summary>
+    public string BundleRegistrationFunctionName { get; set; } = "mono_register_resources_bundle";
+
+    /// <summary>
+    /// The filename for the generated source file that registers
+    /// the bundled resources.
+    /// </summary>
+    public string? BundleFile { get; set; }
+
+    /// <summary>
+    /// Path to store build artifacts
+    /// <summary>
+    public string OutputDirectory {get; set; } = default!;
+
+    /// <summary>
+    /// Resources that were bundled
+    ///
+    /// Successful bundling will set the following metadata on the items:
+    ///   - DataSymbol
+    ///   - LenSymbol
+    /// <summary>
+    [Output]
+    public ITaskItem[] BundledResources { get; set; } = default!;
+
+    public override bool Execute()
+    {
+        List<ITaskItem> bundledResources = new(FilesToBundle.Length);
+        foreach (ITaskItem bundledResource in FilesToBundle)
+        {
+            var resourcePath = bundledResource.ItemSpec;
+
+            bundledResource.SetMetadata("ResourceType", "DataResource");
```

Comment:
We wanted this task to be ran more easily without needing to do too much work on the targets side, so users can pass just a list of resources to bundle instead of needing to perform computations/modify items beforehand for proper usage. I believe making every resource go through this block would help ensure that resources intended to be assemblies aren't corrupted.